File: internal/plant/toxicity.go
package plant

import (
	"slices"
	"time"
)

// Harm is how badly a plant treats whatever ate it, graded by what the owner
// should do rather than by pharmacology, because that is the only scale a
// houseplant app can honestly defend. adr/0003 has the reasoning.
type Harm string

const (
	// The default, and the only honest answer for a plant nobody looked up.
	// Distinct from safe because silence about a lily is not a lily being fine.
	HarmUnknown Harm = "unknown"

	// No toxic principle reported. Which the ASPCA means as "not poisonous",
	// not "harmless": enough of anything causes vomiting.
	HarmSafe Harm = "safe"

	// Self-limiting. Rinse the mouth, offer water, watch it.
	HarmMild Harm = "mild"

	// Systemic effects are plausible. Ring the vet or the poison line.
	HarmModerate Harm = "moderate"

	// Can kill, or has a narrow treatment window. Go now, before signs appear.
	HarmSevere Harm = "severe"
)

// Basis records who did the grading. The ASPCA publishes toxic or non-toxic
// and nothing between, so any finer level is Planty's own judgement and must
// say so rather than borrow the source's authority.
type Basis string

const (
	// The source stated this level itself.
	BasisSource Basis = "source"

	// Graded here from the toxic principle. An honest guess, labelled.
	BasisDerived Basis = "derived"
)

// Known reports whether anyone has actually looked this up.
func (h Harm) Known() bool { return h != "" && h != HarmUnknown }

// Urgent reports whether eating it means calling someone rather than watching.
func (h Harm) Urgent() bool { return h == HarmModerate || h == HarmSevere }

// Toxicity is what a plant does to whatever eats it, rated per audience: an
// Easter lily is renal failure in a cat and a stomach ache in a dog, and one
// flag for both either panics dog owners or kills cats.
type Toxicity struct {
	Cats   Harm `json:"cats"`
	Dogs   Harm `json:"dogs"`
	People Harm `json:"people"`

	Basis Basis `json:"basis,omitempty"`

	// The botanical name this was looked up under. "Lily" is six unrelated
	// plants across three mechanisms, so a rating that cannot say what it
	// resolved to is not evidence of anything.
	IdentifiedAs string `json:"identified_as,omitempty"`

	// The mechanism, in the reference works' vocabulary ("insoluble calcium
	// oxalates"). Justifies the ratings and outlives them when re-checked.
	Principle string `json:"principle,omitempty"`

	// What you would see in an animal that ate it.
	Signs string `json:"signs,omitempty"`

	// Which parts hurt. Structured rather than prose because a tomato's fruit
	// is dinner and its leaves are not, and burying that distinction in a
	// paragraph is how someone gets hurt.
	Parts []string `json:"parts,omitempty"`

	// How it gets you. Euphorbia sap damages eyes without anyone eating it.
	Routes []string `json:"routes,omitempty"`

	// What the ratings cannot carry: why the audiences diverge, dose caveats.
	Notes string `json:"notes,omitempty"`

	// Set only where the advice implied by the rating is actively wrong: a cat
	// that groomed lily pollen needs a vet before any sign appears, and
	// Euphorbia sap in an eye is irrigation rather than anything oral.
	FirstAid string `json:"first_aid,omitempty"`

	// Where this came from, so a doubtful entry can be re-read rather than
	// argued about.
	Source string `json:"source,omitempty"`

	// Separates "looked it up and found nothing" from "never looked", which
	// otherwise wear the same HarmUnknown.
	CheckedAt *time.Time `json:"checked_at,omitempty"`
}

// PartNames is every plant part a hazard can be pinned to.
var PartNames = []string{
	"all", "bulb", "leaf", "stem", "sap", "flower", "fruit", "seed", "root",
}

// RouteNames is every way a plant can reach you.
var RouteNames = []string{"eaten", "skin", "eyes", "breathed"}

// Rating pairs an audience with its verdict.
type Rating struct {
	Who  string
	Harm Harm
}

// Audiences is every rating, in display order. Anything iterating them uses
// this rather than repeating the list.
func (t Toxicity) Audiences() []Rating {
	return []Rating{{"cats", t.Cats}, {"dogs", t.Dogs}, {"people", t.People}}
}

// Checked reports whether anybody has established anything at all.
func (t Toxicity) Checked() bool {
	return t.CheckedAt != nil || t.Cats.Known() || t.Dogs.Known() || t.People.Known()
}

// Worst is the highest rating across the audiences.
func (t Toxicity) Worst() Harm {
	worst := HarmUnknown
	for _, a := range t.Audiences() {
		if rank(a.Harm) > rank(worst) {
			worst = a.Harm
		}
	}
	return worst
}

// Dangerous reports whether this plant is worth warning about for anyone.
func (t Toxicity) Dangerous() bool { return t.Worst().Urgent() }

// Diverges reports whether the audiences disagree. Most of a collection is
// Araceae, which treats all three identically, so the rare plant that does not
// is worth saying out loud rather than leaving someone to spot it in chips.
func (t Toxicity) Diverges() bool {
	return t.Checked() && (t.Cats != t.Dogs || t.Dogs != t.People)
}

// rank orders Harm. Unknown deliberately outranks safe, so sorting by risk
// surfaces the plants nobody has checked instead of burying them.
func rank(h Harm) int {
	switch h {
	case HarmSafe:
		return 1
	case HarmUnknown, "":
		return 2
	case HarmMild:
		return 3
	case HarmModerate:
		return 4
	case HarmSevere:
		return 5
	}
	return 0
}

// ValidHarm reports whether a rating is one the store accepts. The empty
// string counts, because an untouched plant is unrated rather than invalid.
func ValidHarm(h Harm) bool {
	switch h {
	case "", HarmUnknown, HarmSafe, HarmMild, HarmModerate, HarmSevere:
		return true
	}
	return false
}

// Valid holds the two rules that keep this field honest: a claim that
// something is dangerous has to name the mechanism, and a level finer than the
// sources publish has to admit it was derived here.
func (t Toxicity) Valid() error {
	for _, a := range t.Audiences() {
		if !ValidHarm(a.Harm) {
			return invalid("unknown toxicity rating %q for %s", a.Harm, a.Who)
		}
	}
	for _, part := range t.Parts {
		if !slices.Contains(PartNames, part) {
			return invalid("%q is not a plant part; use one of %v", part, PartNames)
		}
	}
	for _, route := range t.Routes {
		if !slices.Contains(RouteNames, route) {
			return invalid("%q is not an exposure route; use one of %v", route, RouteNames)
		}
	}
	switch t.Basis {
	case "", BasisSource, BasisDerived:
	default:
		return invalid("unknown basis %q", t.Basis)
	}
	if t.Worst().Urgent() && t.Principle == "" {
		return invalid("a rating of %q needs a toxic principle to justify it", t.Worst())
	}
	if t.Checked() && t.Basis == "" {
		return invalid("say whether the rating came from the source or was derived here")
	}
	return nil
}

File: ios/Planty/Models/Toxicity.swift
import Foundation

/// How dangerous a plant is to one audience. `unknown` is not a gentler `safe`:
/// it means nobody looked, and every rendering of it has to say exactly that.
enum ToxicityRating: String, FallbackDecodable, CaseIterable {
    case unknown
    case safe
    case mild
    case moderate
    case severe

    static let fallback = ToxicityRating.unknown

    /// Unknown outranks safe on purpose, so the worst rating across a household
    /// can never come back "safe" while one audience was never checked.
    var severityOrder: Int {
        switch self {
        case .safe: 0
        case .unknown: 1
        case .mild: 2
        case .moderate: 3
        case .severe: 4
        }
    }

    /// Somebody established this rating. False for `unknown`, which is what
    /// keeps it out of every calm, filled, reassuring treatment.
    var isChecked: Bool { self != .unknown }

    /// Never "safe", never blank. This label is the whole distance between
    /// "nobody checked" and "probably fine".
    var label: String {
        switch self {
        case .unknown: "Not checked"
        case .safe: "Safe"
        case .mild: "Mild"
        case .moderate: "Moderate"
        case .severe: "Severe"
        }
    }

    var symbol: String {
        switch self {
        case .unknown: "questionmark.circle"
        case .safe: "checkmark.circle.fill"
        case .mild: "exclamationmark.circle.fill"
        case .moderate: "exclamationmark.triangle.fill"
        case .severe: "exclamationmark.octagon.fill"
        }
    }

    /// Lowercase fragment for the divergence sentence, which reads them in a
    /// row: "Severe for cats, mild for dogs and people."
    var clause: String {
        switch self {
        case .unknown: "not checked"
        case .safe: "safe"
        case .mild: "mild"
        case .moderate: "moderate"
        case .severe: "severe"
        }
    }

    func sentence(for audience: ToxicityAudience) -> String {
        switch self {
        case .unknown: "Nobody has checked this for \(audience.plural)."
        case .safe: "Recorded as safe for \(audience.plural)."
        case .mild: "Mildly toxic to \(audience.plural)."
        case .moderate: "Moderately toxic to \(audience.plural)."
        case .severe: "Severely toxic to \(audience.plural)."
        }
    }
}

/// Who is at risk. Three separate answers because they are genuinely different
/// animals, which is the entire reason this data is worth carrying.
enum ToxicityAudience: String, CaseIterable, Sendable, Hashable {
    case cats
    case dogs
    case people

    var plural: String { rawValue }

    var label: String {
        switch self {
        case .cats: "Cats"
        case .dogs: "Dogs"
        case .people: "People"
        }
    }

    var symbol: String {
        switch self {
        case .cats: "cat.fill"
        case .dogs: "dog.fill"
        case .people: "person.fill"
        }
    }
}

/// Where the grading came from. `derived` has to be said out loud: sources
/// publish toxic or not, so anything finer than that is Planty inferring.
enum ToxicityBasis: String, FallbackDecodable, CaseIterable {
    case source
    case derived
    case unknown

    static let fallback = ToxicityBasis.unknown
}

/// Species-level toxicity, as the service sends it. Everything past the three
/// ratings is optional, and a plant nobody looked up carries only unknowns.
struct Toxicity: Codable, Sendable, Hashable {
    var cats: ToxicityRating = .unknown
    var dogs: ToxicityRating = .unknown
    var people: ToxicityRating = .unknown
    var basis: ToxicityBasis = .unknown

    var identifiedAs: String?
    var principle: String?
    var signs: String?
    var parts: [String] = []
    var routes: [String] = []
    var notes: String?
    var firstAid: String?
    var source: String?
    var checkedAt: Date?

    enum CodingKeys: String, CodingKey {
        case cats
        case dogs
        case people
        case basis
        case identifiedAs = "identified_as"
        case principle
        case signs
        case parts
        case routes
        case notes
        case firstAid = "first_aid"
        case source
        case checkedAt = "checked_at"
    }
}

extension Toxicity {
    /// An absent rating decodes to `unknown`, never to nothing-and-so-fine.
    /// A rating the app has never heard of does the same rather than guessing.
    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        cats = try container.decodeIfPresent(ToxicityRating.self, forKey: .cats) ?? .unknown
        dogs = try container.decodeIfPresent(ToxicityRating.self, forKey: .dogs) ?? .unknown
        people = try container.decodeIfPresent(ToxicityRating.self, forKey: .people) ?? .unknown
        basis = try container.decodeIfPresent(ToxicityBasis.self, forKey: .basis) ?? .unknown

        identifiedAs = try container.decodeIfPresent(String.self, forKey: .identifiedAs)
        principle = try container.decodeIfPresent(String.self, forKey: .principle)
        signs = try container.decodeIfPresent(String.self, forKey: .signs)
        parts = try container.decodeIfPresent([String].self, forKey: .parts) ?? []
        routes = try container.decodeIfPresent([String].self, forKey: .routes) ?? []
        notes = try container.decodeIfPresent(String.self, forKey: .notes)
        firstAid = try container.decodeIfPresent(String.self, forKey: .firstAid)
        source = try container.decodeIfPresent(String.self, forKey: .source)
        checkedAt = try container.decodeIfPresent(Date.self, forKey: .checkedAt)
    }
}

extension Toxicity {
    func rating(for audience: ToxicityAudience) -> ToxicityRating {
        switch audience {
        case .cats: cats
        case .dogs: dogs
        case .people: people
        }
    }

    var ratings: [(audience: ToxicityAudience, rating: ToxicityRating)] {
        ToxicityAudience.allCases.map { ($0, rating(for: $0)) }
    }

    /// The worst of the three, with unknown ranked above safe, so a card can
    /// never headline "safe" on the strength of the columns that were filled in.
    var worst: ToxicityRating {
        ratings.map(\.rating).max { $0.severityOrder < $1.severityOrder } ?? .unknown
    }

    /// The three do not agree. Rare, and precisely the case that kills animals:
    /// a lily is a sore stomach for a dog and renal failure for a cat.
    var diverges: Bool { Set(ratings.map(\.rating)).count > 1 }

    /// Nobody has looked this up at all.
    var isUnchecked: Bool {
        checkedAt == nil && ratings.allSatisfy { !$0.rating.isChecked }
    }

    /// Somebody looked and still could not say, which is a different sentence
    /// from nobody having looked.
    var isCheckedButUnresolved: Bool {
        checkedAt != nil && ratings.allSatisfy { !$0.rating.isChecked }
    }

    /// Planty graded this itself rather than repeating what the source said.
    var isDerived: Bool { basis == .derived }

    /// One line, worst first, so nobody has to diff three chips to find out
    /// that something in this pot can kill a cat.
    var headline: String {
        switch worst {
        case .severe: "Severely toxic"
        case .moderate: "Moderately toxic"
        case .mild: "Mildly toxic"
        case .unknown: "Not checked"
        case .safe: "No known toxicity"
        }
    }

    /// Spelled out in words when the audiences disagree, because a difference
    /// nobody notices is the same as no data at all.
    var divergenceSentence: String? {
        guard diverges else { return nil }
        let byRating = Dictionary(grouping: ratings, by: \.rating)
        let clauses = byRating.keys
            .sorted { $0.severityOrder > $1.severityOrder }
            .map { rating in
                let names = (byRating[rating] ?? []).map(\.audience.plural)
                return "\(rating.clause) for \(Self.listed(names))"
            }
        let sentence = clauses.joined(separator: ", ") + "."
        return sentence.prefix(1).uppercased() + sentence.dropFirst()
    }

    func checkedLine(now: Date = Date()) -> String? {
        guard let checkedAt else { return nil }
        return "Checked \(RelativeAge.phrase(since: checkedAt, now: now))."
    }

    private static func listed(_ names: [String]) -> String {
        switch names.count {
        case 0: ""
        case 1: names[0]
        case 2: "\(names[0]) and \(names[1])"
        default: "\(names.dropLast().joined(separator: ", ")) and \(names[names.count - 1])"
        }
    }
}
